discojs/src/processing/index.ts

```
@@ -0,0 +1,172 @@
+      const maxLength =
+        t.trainingInformation.maxSequenceLength ??
+        (tokenizer.model_max_length as number) + 1;
+
+      return d
+        .map((line) =>
+          processing.tokenizeAndLeftPad(line, tokenizer, maxLength),
+        )
+        .map((tokens) => tokens.pop());
+    }
+  }
+}
+
+export async function postprocess<D extends DataType>(
+  task: Task<D>,
+  dataset: Dataset<ModelEncoded[D][1]>,
+): Promise<Dataset<Inferred[D]>> {
+  switch (task.trainingInformation.dataType) {
+    case "image": {
+      // cast as typescript doesn't reduce generic type
+      const d = dataset as Dataset<ModelEncoded["image"][1]>;
+      const { LABEL_LIST } =
+        task.trainingInformation as TrainingInformation<"image">;
+      const labels = List(LABEL_LIST);
+
+      return d.map((index) => {
+        const v = labels.get(index);
+        if (v === undefined) throw new Error("index not found in labels");
+        return v;
+      }) as Dataset<Inferred[D]>;
+    }
+    case "tabular": {
+      // cast as typescript doesn't reduce generic type
+      const d = dataset as Dataset<ModelEncoded["tabular"][1]>;
+      const { outputColumn } =
+        task.trainingInformation as TrainingInformation<"tabular">;
+
+      return d.map((row) =>
+        Object.fromEntries([[outputColumn, row]]),
+      ) as Dataset<Inferred[D]>;
+    }
+    case "text": {
+      // cast as typescript doesn't reduce generic type
+      const d = dataset as Dataset<ModelEncoded["text"][1]>;
+      const t = task as Task<"text">;
+      const tokenizer = await models.getTaskTokenizer(t);
+
+      return d.map((token) => tokenizer.decode([token])) as Dataset<
+        Inferred[D]
+      >;
+    }
+  }
+}
+
+function extractToNumbers(columns: Iterable<string>, row: Tabular) {
+  return (
+    List(columns)
+      .map((column) => processing.extractColumn(row, column))
+      // TODO sanitization doesn't care about column distribution
+      .map((v) => (v !== "" ? v : "0"))
```

I chose a constant value as a temporary fix for the training failures (#615, #649) but now that the preprocessing works differently we may be able to handle it more gracefully. 
IMO choosing how to handle missing values should be the responsibility of the data owner so we could throw an error or drop the whole row as default behavior.

---

good point, thanks for linking the issue; followup on #649